### ISSUE TITLE
Differences in counting testcases on LAVA vs SQUAD side

### ISSUE THREAD

**Issue by pfalcon** [Linaro/squad#925]
Steps to reproduce:

1. Go to https://qa-reports.linaro.org/lite/lite-tf-m/build/b157dca4-130/ .
2. Due to #924, you can't see number of testcases in that build, so uncheck the "failures only" checkbox, and behold the figure "108".
3. Click on the "job_url" link presented on that page, which is https://validation.linaro.org/scheduler/job/2246506
4. Click on "Result" icon.
5. Behold that the "tfm-tests" testsuite has 176 testcase total (and pass).

So, the issue here is that some testcase names are duplicate. Which is very sad situation, and it shouldn't exist. And it exists, and when it happens, SQUAD vs LAVA shouldn't show discrepancies in numbers, and apparently LAVA's counting is authoritative.

Regardless of whether we workaround that on our side (I'm not sure if it can be worked around in the general case), it would be very nice to have this fixed in SQUAD. There would be at least following ways to address it:

1. Accept, and thus allow, that testcase names can be duplicate. As far as I see, LAVA does export "testcase IDs", so there should not a problem for SQUAD to identify a particular testcase.
2. If it's now too deep in SQUAD architecture that testcase name is a primary ID of a testcase, then when importing from LAVA, anticipate the duplicates, and e.g. suffix the name with an incrementing counter.

Thanks.

**Comment by mwasilew** [comment]
> Accept, and thus allow, that testcase names can be duplicate. As far as I see, LAVA does export "testcase IDs", so there should not a problem for SQUAD to identify a particular testcase.

This is a unique ID from LAVA database. Not very useful for SQUAD.

> If it's now too deep in SQUAD architecture that testcase name is a primary ID of a testcase, then when importing from LAVA, anticipate the duplicates, and e.g. suffix the name with an incrementing counter.

Test case name is not a primary ID in SQUAD. However adding suffixes is of little use as well.

IMHO you have a problem with either test case naming or test case parsing in LAVA. There is no way to meaningfully identify test cases with the same names. Let's assume that you send 2 test cases to SQUAD. Both are named 'abc'. They will be processed in SQUAD in random order. What you propose is to add _1 and _2 to the names so we'll end up with abc_1 and abc_2. So far so good. We have unique names. This works as long as both tests give the same result. If first instance of abs is pass and second instance of abc is fail, you will get a random regression. There is no guarantee that abc_1 from one lava job will be the same abc_1 in the next lava job. They can, but don't have to. So the only way to fix it is to change the test names in your suite.

Luckily there are unique names in the suite. As example, there is 'get_interface_with_valid_data' which appears twice. It corresponds to 'TFM_ITS_TEST_2005' and 'TFM_SST_TEST_1005'. Just use these IDs and your problem is solved.

**Comment by pfalcon** [comment]
Thanks for the detailed response.

> Not very useful for SQUAD.

As was mentioned, useful at least to not display to a user a different number of testcases than LAVA does.

> So the only way to fix it is to change the test names in your suite.

But it's not my suite. It somebody else's suite.

> If first instance of abs is pass and second instance of abc is fail, you will get a random regression.

That will be sad indeed. However currently, there's even bigger regression with the number of tests displayed in LAVA vs SQUAD.

> Luckily there are unique names in the suite. As example, there is 'get_interface_with_valid_data' which appears twice. It corresponds to 'TFM_ITS_TEST_2005' and 'TFM_SST_TEST_1005'. Just use these IDs and your problem is solved.

No, it's not solved. It's worked around. Because names like "TFM_ITS_TEST_2005" are *absolutely* useless to someone not familiar with the nature of tests performed, and *rather* useless even to someone who know them, because they will need to dig deep into source to even understand what's failed, instead of potentially get an immediate insight what's wrong, and maybe go straight to fix it in the "functional" source code, instead of wading thru the testsuite source. After all, the primary purpose of a name is *being descriptive* (there're IDs for anything else), and that's why I was keen to use the descriptive testcase name.


Generally, when I consider whether to silently workaround an issue, or report it, I try to remember that LAVA, SQUAD, other Linaro software are intended to be ultimately used and usable to our members. So, I err on the side of trying to improve it, rather than not, and doing my small part by at least reporting issues I've seen. Of course, I may be wrong with any particular case.

Thanks again.

**Comment by pfalcon** [comment]
Even myself, not much familiar with the TF-M, would get an idea that it's a kind of "matrix" test, where the same function is exercised under different conditions, so the culprit is likely in get_interface functionality.

But that's not the point, Milosz. The point is that different number of testcases are displayed wrt to LAVA. That's the real issue.

(And if you say that it's such a corner case, and that you have so many more important things to take care of, and that's why you closed it, I would just accept it as is.)

**Comment by mwasilew** [comment]
Android can successfully name 'matrix' tests with the corresponding variant (for locale tests). I'm sure TF-M can do the same. I guess you didn't read my previous reply. LAVA doesn't care what the test case name is. It works in a single job context. SQUAD does care what the name is as it's used as basis for test comparison between versions (builds). As I explained, there is no way of reliably determining which test fails in case more than one carry the same name. So there is no bug in SQUAD. I would also argue there is no bug in LAVA. You're just using it wrong. Please note there is no 'get_interface_with_valid_data' (lava converts it to lowercase and replaces whitespaces with _)anywhere in the TF-M source code, while you probably can find 'TFM_SST_TEST_1005'. So the argument that the 'English' name of the test is better isn't convincing.

My opinion (which can be disregarded) is that the fix should be done in TF-M. Most importantly it should allow to remove colour codes from the output. Once this is done you could try to make the output for a single test in one line. This would allow you to create easier to read regex and solve the 'problem' you're describing.

**Comment by mwasilew** [comment]
and you can do sth like this: https://validation.linaro.org/scheduler/job/2248566

**Comment by pfalcon** [comment]
> So there is no bug in SQUAD. I would also argue there is no bug in LAVA. You're just using it wrong.

I just hope that I won't get caught in a need to convey that to our members or otherwise "customers". My blushing would give away that I myself don't believe in such answers ;-).

> So the argument that the 'English' name of the test is better isn't convincing.

My only argument was that there should be not discrepancy in testcase numbers between LAVA and SQUAD. Beyond that, I tried to draft a couple of ways to address that. If you don't agree with the original problem statement, then you certainly won't agree with any ways I sketched to address them, seeing only drawbacks in them.

> My opinion (which can be disregarded) is that the fix should be done in TF-M.

On an occasion, I'll let them know that SQUAD thinks that the fix should be on their side. But we won't be able to patch every project out there. We either have a flexible, general-purpose testing system which can work with the real world, or ... we don't. I also think that when you suggest that, you imagine either patching, or writing separate test helper in a high-level language, like shell, which we can just super-impose on top of an existing system (i.e. a typical Linux-like system). That's not what we deal with in LITE. We have monolithic, atomic binaries, and tests are raw C files somewhere deep in the original project. To change the tests (assuming they don't want), we effectively need to fork the entire project. And then when we do that, there's a concern that we no longer test upstream (because composability of those systems is again isn't comparable to that of a typical Linux system).

> and you can do sth like this: https://validation.linaro.org/scheduler/job/2248566

So, "ideal" solution would indeed be to use a combo of "id" and "description", it's just LAVA is not expressive enough to do that (would need a post-processing step on the captured `?P<test_case_id>`, thought about submitting a LAVA feature request for that, but figured it has much lower chance to acted upon, so kept that to myself). Who'd new that newline characters are so cutely being taken care of, as you show! And that "description" isn't that much of a sleeve in the eye, as one could think speculatively. I also had idea to split the test results in separate suites, though got stuck on this perplexing 179 vs 108 testcases discrepancy.

So, all are the good hints, that's what I'm going to do I guess. And thanks for the discussion again!

**Comment by pfalcon** [comment]
> and you can do sth like this: https://validation.linaro.org/scheduler/job/2248566

Ok, I tried that pattern, with global matching across the entire TF-M testsuite output. And got 158 testcases in the LAVA itself! (vs 174 with were gotten originally, sorry, above 179 is used, it's 174 + 5, where 5 are internel "lava" testsuite's "testcases".)

Debugging what's wrong where, I got an explicit confirmation of an intuitive hunch I maintained for a while: it's dangerous (in a sense of getting inadequate results) to use greedy regex repetitions **anywhere** in the match patterns. Just imagine yourself: if you wrote a greed regex which matches within single line, then if applied to 2 lines, it'll happily match across them. And it's all not even deterministic, as it depends in how much "expect" slurped into its buffer (that's of course hardware and cosmic rays dependent). expect's matching across newlines is definitely powerful feature, which allows to do things we discuss here (== useful), but if not accounted for, can lead to a mess (and false negatives, or worse, positives). Indeed, here I was getting testcase names like:

`tfm_its_test_2003_description_set_interface_with_null_data_pointer_32m_test_passed_37m_executing_tfm_its_test_2004_description_set_interface_with_write_once_uid`

The bottom line: never use standalone `*` and `+` in the LAVA matching regexes, it always should be `*?` and `+?`.

Another issue is that I never saw such a warning or recommendation in the LAVA docs. Nor can I say that I saw it much (if at all) the test definitions I eyeballed. If anything, my original linked job for this issue used `+?` (because as I said, I maintained the suspicion for a while, even though never tried to "catch" it by explicit reproduction), and then @mwasilew's sample job used just `+` again.


(I'm writing all this in such a detail, to server as my own reference, going to link it from my LAVA notes, and to be able to refer other people to.)

**Comment by mwasilew** [comment]
> I just hope that I won't get caught in a need to convey that to our members or otherwise "customers". My blushing would give away that I myself don't believe in such answers ;-).

So you rather spend weeks of engineering time in Linaro than propose a few lines patch in 
https://git.trustedfirmware.org/TF-M/tf-m-tests.git/tree/test/framework/test_framework.c
? It's really not that hard to add p_test->name in line 131 and passing p_test to test_failed so p_test-> name can also be displayed in TEST_LOG("%s  TEST FAILED!\r\n", p_test->name); (this should go to line 29)?

I also think your assumption about 'English readable' test names is incorrect. If you used code names that come from tf-m-tests your problem would had been solved already.

**Comment by mwasilew** [comment]
@pfalcon just FYI: https://review.trustedfirmware.org/c/TF-M/tf-m-tests/+/7370. It's not impossible to get your changes merged in TF code.

**Comment by pfalcon** [comment]
@mwasilew: Cool, thanks. But I didn't challenge that. I chellange an idea that it's practical to change every other project out there. Vs an alternative idea of having a test system which just can work with any project out there.

**Comment by mwasilew** [comment]
With enough budget your idea might work. Unfortunately I don't have the budget for it.

**Comment by pfalcon** [comment]
I said that I could anticipate, and understand, this reasoning yet in https://github.com/Linaro/squad/issues/925#issuecomment-738732738 comment.

This is pretty minor issue anyway (as easy workarounds exists). I'd say we've bikeshedded it enough by now ;-). (And why I did more-than-it's-probably-worth discussion on it, exactly because it seemed so simple. It's different matter when issue is bigger, and how to solve it is less clear. Again, you see deeper and wider into this stuff, so if you say it's not worth it, then it's just that way. Thanks.)
